src/libos/crates/sgx-disk/src/host_disk/sync_io_disk.rs

```diff
@@ -104,6 +77,93 @@ impl SyncIoDisk {
         let end_offset = end_block * BLOCK_SIZE;
         Ok((begin_offset, end_offset))
     }
+
+    fn preadv(&self, fd: i32, bufs: &mut [BlockBuf], offset: i64) -> Result<usize> {
+        let mut total_len = 0;
+
+        let iovecs: Box<Vec<libc::iovec>> = Box::new(
+            bufs.iter_mut()
+                .map(|buf| {
+                    let iov_base = buf.as_slice_mut().as_mut_ptr();
+                    let iov_len = buf.as_slice().len();
+                    total_len += iov_len;
+                    libc::iovec {
+                        iov_base: iov_base as _,
+                        iov_len,
+                    }
+                })
+                .collect(),
+        );
+        let (iovecs_ptr, iovecs_len) =
+            { (iovecs.as_ptr() as *const libc::iovec, iovecs.len() as i32) };
+
+        // TODO: fix this limitation
+        const LINUX_IOVS_MAX: i32 = 1024;
+        debug_assert!(iovecs_len < LINUX_IOVS_MAX);
+
+        let read_len = unsafe {
+            #[cfg(not(feature = "sgx"))]
+            let read_len = libc::preadv64(fd, iovecs_ptr, iovecs_len, offset);
+            #[cfg(feature = "sgx")]
+            let read_len = libc::ocall::preadv64(fd, iovecs_ptr, iovecs_len, offset);
+            read_len
+        };
+
+        if read_len < 0 {
+            return_errno!(Errno::from(libc_errno() as u32), "libc::preadv64 error");
+        }
+        // Partial reads are unlikely to happen since buffers and offset are all within legal range
+        // of the host file (file size is truncated to size of total blocks when created)
+        assert!(
+            read_len == total_len as isize,
+            "libc::preadv64 return wrong read length"
+        );
+
+        Ok(read_len as usize)
+    }
+
+    fn pwritev(&self, fd: i32, bufs: &[BlockBuf], offset: i64) -> Result<usize> {
+        let mut total_len = 0;
+
+        let iovecs: Box<Vec<libc::iovec>> = Box::new(
+            bufs.iter()
+                .map(|buf| {
+                    let iov_base = buf.as_slice().as_ptr();
+                    let iov_len = buf.as_slice().len();
+                    total_len += iov_len;
+                    libc::iovec {
+                        iov_base: iov_base as _,
+                        iov_len,
+                    }
+                })
+                .collect(),
+        );
+        let (iovecs_ptr, iovecs_len) =
+            { (iovecs.as_ptr() as *const libc::iovec, iovecs.len() as i32) };
+
+        // TODO: fix this limitation
+        const LINUX_IOVS_MAX: i32 = 1024;
+        debug_assert!(iovecs_len < LINUX_IOVS_MAX);
```

Comment:
My second though is that we should remove this assert.

The man page of `pwritev` states.

> EINVAL
> The sum of the iov_len values overflows an ssize_t value. Or, the vector count iovcnt is less than zero or greater than the permitted maximum.

> Linux notes
> POSIX.1-2001 allows an implementation to place a limit on the number of items that can be passed in iov. An implementation can advertise its limit by defining IOV_MAX in <[limits.h](https://linux.die.net/include/limits.h)> or at run time via the return value from sysconf(_SC_IOV_MAX). On Linux, the limit advertised by these mechanisms is 1024, which is the true kernel limit. However, the glibc wrapper functions do some extra work if they detect that the underlying kernel system call failed because this limit was exceeded. In the case of readv() the wrapper function allocates a temporary buffer large enough for all of the items specified by iov, passes that buffer in a call to [read](https://linux.die.net/man/2/read)(2), copies data from the buffer to the locations specified by the iov_base fields of the elements of iov, and then frees the buffer. The wrapper function for writev() performs the analogous task using a temporary buffer and a call to [write](https://linux.die.net/man/2/write)(2).

Linux returns `EINVAL` if the length of iovecs is too long. Or the glibc wrapper can combines the iovecs into a big buffer. As this error can be captured and returned to the user at runtime, we should give an opportunity to handle it instead of panic.

Reply:
Make sense to me. Done